data-prepper-api/src/main/java/com/amazon/dataprepper/model/buffer/AbstractBuffer.java

```diff
@@ -76,10 +80,18 @@ public void write(T record, int timeoutInMillis) throws TimeoutException {
     @Override
     public Map.Entry<Collection<T>, CheckpointState> read(int timeoutInMillis) {
         final Map.Entry<Collection<T>, CheckpointState> readResult = readTimer.record(() -> doRead(timeoutInMillis));
-        recordsReadCounter.increment(readResult.getKey().size()*1.0);
+        recordsInflightCounter.increment(readResult.getValue().getNumRecordsToBeChecked()*1.0);
         return readResult;
     }
 
+    @Override
+    public void checkpoint(final CheckpointState checkpointState) {
+        checkpointTimer.record(() -> doCheckpoint(checkpointState));
+        final int numRecordsToBeChecked = checkpointState.getNumRecordsToBeChecked();
+        recordsInflightCounter.increment(-numRecordsToBeChecked);
```

Comment:
counters can only go up: https://micrometer.io/docs/concepts#_counters. while this may work in a unit test because we are using a SimpleRegistry which keeps the latest value in memory only, this will not work in production.

Reply:
this will need to be a gauge